### PR TITLE
replace `_CCCL_ALWAYS_INLINE` with `_CCCL_FORCEINLINE`

### LINKED ISSUE #2438: [BUG]: unify `_CCCL_FORCEINLINE` and `_CCCL_ALWAYS_INLINE`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Something else

### Component

libcu++

### Describe the bug

cccl has `_CCCL_FORCEINLINE` and `_CCCL_ALWAYS_INLINE`. there should be only one. also, `_CCCL_FORCEINLINE` currently expands to `inline` when not using a CUDA compiler. that is unexpected. it should expand to either `__attribute__((always_inline))` or `__forceinline` depending on which is supported by the host compiler.

### How to Reproduce

n/a

### Expected behavior

n/a

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by ericniebler**
## Description
cccl has `_CCCL_FORCEINLINE` and `_CCCL_ALWAYS_INLINE`. there should be only one. also, `_CCCL_FORCEINLINE` currently expands to `inline` when not using a CUDA compiler. that is unexpected. it should expand to either `__attribute__((always_inline))` or `__forceinline` depending on which is supported by the host compiler.

closes #2438 

This PR moves the definition of `_CCCL_FORCEINLINE` from `execution_space.h` to `visibility.h`. it also changes the definition to expand directly to either `__inline__ __attribute__((always_inline))` or `__forceinline` rather then indirectly through the `__forceinline__` macro defined in `host_defines.h`.

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by miscco — APPROVED** [review]

File: thrust/testing/transform_input_output_iterator.cu [hunk 1]
```diff
@@ -7,8 +7,16 @@
...
+// Adding this attribute suffices that this miscompilation does not appear anymore
+#if defined(_CCCL_COMPILER_GCC) && __GNUC__ >= 11
+#  define THRUST_DISABLE_BROKEN_GCC_VECTORIZER __attribute__((optimize("no-tree-vectorize")))
+#else
+#  define THRUST_DISABLE_BROKEN_GCC_VECTORIZER
+#endif
+
 template <class Vector>
-void TestTransformInputOutputIterator()
+THRUST_DISABLE_BROKEN_GCC_VECTORIZER void TestTransformInputOutputIterator()
```
Comment (ericniebler) [inline thrust/testing/transform_input_output_iterator.cu:19]: This fixes our tests, but won't gcc still be miscompiling Thrust for users?

Comment (miscco) [reply thrust/testing/transform_input_output_iterator.cu:19]: That is nothing we can change. I want to note that this is exceptionally frickle and dependent on exact sizes and otimization settings, so I dont see anything we can do there

File: thrust/thrust/detail/function.h [hunk 2]
```diff
@@ -39,7 +39,7 @@ struct wrapped_function
 
   _CCCL_EXEC_CHECK_DISABLE
   template <typename... Ts>
-  _CCCL_FORCEINLINE _CCCL_HOST_DEVICE Result operator()(Ts&&... args) const
+  inline _CCCL_HOST_DEVICE Result operator()(Ts&&... args) const
```
Comment (ericniebler) [inline thrust/thrust/detail/function.h:42]: @miscco at least locally, this change avoids the gcc optimizer issue.

Comment (miscco) [reply thrust/thrust/detail/function.h:42]: That is awesome. I will test whether we can avoid that hack in the other tests too. Will file a separate PR though

Comment (miscco) [reply thrust/thrust/detail/function.h:42]: looks like this is still failing with gcc 12 😿 
https://github.com/NVIDIA/cccl/actions/runs/10967082905/job/30507674109?pr=2439

Comment (ericniebler) [reply thrust/thrust/detail/function.h:42]: it fixed almost all of the failures tho. just two remain. i can track those down when i get back from vacation.

**Pull request merged**